src/weights.jl

```diff
@@ -35,7 +36,33 @@ end
     W(v, sum(v))
 end
 
-Base.getindex(wv::W, ::Colon) where {W <: AbstractWeights} = W(copy(wv.values), sum(wv))
+Base.getindex(wv::AbstractWeights, ::Colon) = copy(wv)
+
+Base.copy(wv::W) where {W <: AbstractWeights} =
+    weightstype(W)(copy(wv.values), sum(wv))
+
+@propagate_inbounds function Base.view(wv::W, inds...) where {S <: Real, W <: AbstractWeights{S}}
+    @boundscheck checkbounds(wv, inds...)
+    @inbounds v = invoke(view, Tuple{AbstractArray, Vararg{Any}}, wv, inds...)
+    # Sum is not actually used but compute the right type for clarity
+    weightstype(W)(v, zero(S))
+end
+
+# This method is implemented for backward compatibility
+@propagate_inbounds function Base.view(wv::W, inds::Union{Integer, CartesianIndex}...) where
+    {S <: Real, W <: AbstractWeights{S}}
+    @boundscheck checkbounds(wv, inds...)
+    @inbounds invoke(view, Tuple{AbstractArray, Vararg{Any}}, wv, inds...)
+end
+
+# Always recompute the sum for views of AbstractWeights, as we cannot know whether
+# the parent array has been mutated
+Base.sum(wv::AbstractWeights{S, T, <:SubArray{T, <:Any, <:AbstractWeights}}) where
+    {S<:Real, T<:Real} = sum(wv.values)
+
+Base.copy(wv::W) where
+    {S<:Real, T<:Real, W<:AbstractWeights{S, T, <:SubArray{T, <:Any, <:AbstractWeights}}} =
+    weightstype(W)(copy(view(parent(wv.values).values, parentindices(wv.values)...)), wv.sum)
```

Comment:
for example here the call seems to be incorrect. You either should call `sum(wv)` or not pass the sum at all and force its computation in the constructor. Am I missing something here?

Reply:
The idea is just to pass a `zero` of the right type (as it determines the return type). I could use `zero(typeof(wv.sum))` for clarity if you prefer, or add a comment.

Reply:
But that is a point that I think it is incorrect, the reason is that when you write `copy(view(parent(wv.values).values, parentindices(wv.values)...))` it is not a view any more (`copy` is applied as the outermost function) so it must have a correct `sum` supplied. At least this is what I understand should happen.

Reply:
Woops, right. So that also needs to be tested...

Reply:
Yes + the point is that this kind of things is tricky to detect when you are having a special treatment of `sum` field. Maybe we should also add a comment in the docs (or at least in the source code) that this is a private field?

Reply:
Fixed and tests added.